File: backend/services/soil.py
"""
SoilGrids REST v2 client — fetches soil properties for a lat/lng point.
No API key required.
"""
import httpx
from models import SoilData

_BASE = "https://rest.isric.org/soilgrids/v2.0/properties/query"

# SoilGrids property names → our field names
_PROP_MAP = {
    "phh2o": "ph",
    "oc": "organic_carbon",
    "nitrogen": "nitrogen",
    "clay": "clay",
    "sand": "sand",
    "silt": "silt",
    "bdod": "bulk_density",
    "cec": "cation_exchange",
}


async def fetch_soil(lat: float, lng: float) -> SoilData:
    params = {
        "lon": lng,
        "lat": lat,
        "property": list(_PROP_MAP.keys()),
        "depth": ["0-5cm"],
        "value": ["mean"],
    }

    values: dict[str, float | None] = {v: None for v in _PROP_MAP.values()}

    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.get(_BASE, params=params)
            resp.raise_for_status()
            data = resp.json()

        for layer in data.get("properties", {}).get("layers", []):
            prop_name = layer.get("name")
            field_name = _PROP_MAP.get(prop_name)
            if not field_name:
                continue
            depths = layer.get("depths", [])
            if depths:
                raw = depths[0].get("values", {}).get("mean")
                if raw is not None:
                    # SoilGrids returns scaled integers; divide by the stated factor
                    factor = layer.get("unit_measure", {}).get("d_factor", 1) or 1
                    values[field_name] = raw / factor
    except Exception:
        pass  # Return nulls on any failure — caller handles gracefully

    return SoilData(source="SoilGrids v2 (ISRIC)", **values)

File: backend/src/services/soil.ts
import { SoilData } from "../models/types";

const BASE = "https://rest.isric.org/soilgrids/v2.0/properties/query";

const PROP_MAP: Record<string, keyof SoilData> = {
  phh2o: "ph",
  oc: "organic_carbon",
  nitrogen: "nitrogen",
  clay: "clay",
  sand: "sand",
  silt: "silt",
  bdod: "bulk_density",
  cec: "cation_exchange",
};

export async function fetchSoil(lat: number, lng: number): Promise<SoilData> {
  const result: any = {
    ph: null,
    organic_carbon: null,
    nitrogen: null,
    clay: null,
    sand: null,
    silt: null,
    bulk_density: null,
    cation_exchange: null,
    source: "SoilGrids v2 (ISRIC)",
  };

  try {
    const url = new URL(BASE);
    url.searchParams.append("lon", lng.toString());
    url.searchParams.append("lat", lat.toString());
    Object.keys(PROP_MAP).forEach((p) => url.searchParams.append("property", p));
    url.searchParams.append("depth", "0-5cm");
    url.searchParams.append("value", "mean");

    const resp = await fetch(url.toString());
    if (!resp.ok) throw new Error(`SoilGrids failed: ${resp.status}`);
    const data: any = await resp.json();

    const layers = data?.properties?.layers || [];
    for (const layer of layers) {
      const propName = layer.name;
      const fieldName = PROP_MAP[propName];
      if (!fieldName) continue;

      const depths = layer.depths || [];
      if (depths.length > 0) {
        const raw = depths[0].values?.mean;
        if (raw !== undefined && raw !== null) {
          const factor = layer.unit_measure?.d_factor || 1;
          result[fieldName] = raw / factor;
        }
      }
    }
  } catch (error) {
    console.error("fetchSoil error:", error);
  }

  return result as SoilData;
}
